Python/chapter01/1.2 - Check Perm/miguel_1.2_sol.py

```diff
@@ -0,0 +1,83 @@
+"""
+Python version 3.7.0
+1.2 - Check Permutation
+Given two strings, write a method to decide if one is a permutation of the other.
+"""
+import unittest
+from typing import Callable
+
+
+def check_permutation(s1: str, s2: str) -> bool:
+    """
+    Determines if a s1 is a permutation of s2.
+    s1 is a permutation of s2 if the characters from s1 map to
+    the same characters in s2 in a one-to-one correspondence AND s1 and s2 are of the same length.
+    Runtime:  O(n + m), where n is the length of s1 and m is the length of s2.
+    Given:		            Expect:
+    "tacos", "costa"            True
+    "tacoos", "costa"           False
+    "swag", "hello"         False
+    "hi", "hello"           False
+    "hola", "loha"          True
+    "p3nding", "dingp3n"            True
+    "", "test"          False
+    "", ""          True
+    "test2", ""         False
+    :param s1: string of size n
+    :param s2: string of size m
+    :return: True if s1 is a permutation of s2, False otherwise
+    """
+    # permutations must be of equal length, one-to-one
+    if len(s1) != len(s2):
+        return False
+    freqs_s1 = {}
+    # build histogram of seen characters in s1
+    # using histogram because string could have repeated characters
+    for c in s1:
+        if c not in freqs_s1:
```

Comment:
this check-if-exists-otherwise-mutate pattern can be avoided with [`collections.defaultdict(int)`](https://docs.python.org/3/library/collections.html#collections.defaultdict)

Reply:
I tried out `collections.defaultdict(int)` and I see the advantage of it, but ended up sticking with using solely `collections.Counter()` objects, as you recommended.